### PR TITLE
release: v13.0.0

### PR DESCRIPTION
**Pull request by georglauterbach**
# Description

This PR marks the feature freeze phase for the `v13.0.0` release.

The current `:edge` image will be tested for roughly **7 days** and if no bugs get reported, `v13.0.0` will be released. During this testing period, only PRs that do not change functionality will be merged.

**If you want to support us, give the latest `:edge` image a try and report any issues you encounter back to this PR.**

Documentation updates may be merged. **Feature freeze is in place until 24 Nov 2023**.

---

@polarathene @casperklein please apply suggestions immediately to the branch, no need to ask me if you want to change stuff up; already spent too much time on `CHANGELOG.md` and drafting the release

## Type of change

- [x] Release

## Release-Checklist:

- [x] Update CHANGELOG.md
- [x] Update version in VERSION
- [ ] Draft release (still needs proper text description for changes)

### CODE REVIEW

**Review by casperklein — APPROVED** [review]

**Comment by IIIlllIIllIIIll** [comment]
It seems that this line of document needs to be changed after #3335.
https://github.com/docker-mailserver/docker-mailserver/blob/d8ebf591f9ff320821c5779521fed1505f1b2916/docs/content/config/advanced/mail-sieve.md?plain=1#L16

**Comment by polarathene** [comment]
> It seems that this line of document needs to be changed after #3335.

Related, not likely to be applicable to LDAP users. Although new setups may want to adapt config for that too.

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,11 +28,87 @@ All notable changes to this project will be documented in this file. The format
...
+  - improve docs about how to work with logs ([#3626](https://github.com/docker-mailserver/docker-mailserver/pull/3626))
+  - update Rspamd directory name ([#3629](https://github.com/docker-mailserver/docker-mailserver/pull/3629))
+  - clarify default for ENV `FETCHMAIL_PARALLEL` ([#3603](https://github.com/docker-mailserver/docker-mailserver/pull/3603))
+  - fix spelling & remove dedicated AllContributors section ([#3638](https://github.com/docker-mailserver/docker-mailserver/pull/3638))
+  - correct path for logs in docs ([#3640](https://github.com/docker-mailserver/docker-mailserver/pull/3640))
+- tests
+  - refactor LDAP tests to current conventions ([#3483](https://github.com/docker-mailserver/docker-mailserver/pull/3483))
+  - change OpenLDAP image to `bitnami/openldap` ([#3494](https://github.com/docker-mailserver/docker-mailserver/pull/3494))
+  - revise LDAP config + setup ([#3514](https://github.com/docker-mailserver/docker-mailserver/pull/3514))
+  - add tests for helper function `_add_to_or_update_postfix_main()` ([#3505](https://github.com/docker-mailserver/docker-mailserver/pull/3505))
```
Comment (polarathene) [inline CHANGELOG.md:89]:
```suggestion
- **Documentation:**
  - Restored missing edit button ([#3338](https://github.com/docker-mailserver/docker-mailserver/pull/3338))
  - Complete rewrite of the IPv6 page ([#3244](https://github.com/docker-mailserver/docker-mailserver/pull/3244))
  - Removed port 25 from FAQ entry for mail client ports supporting authenticated submission ([#3496](https://github.com/docker-mailserver/docker-mailserver/pull/3496))
  - Fixed IPv6 example for Compose ([#3531](https://github.com/docker-mailserver/docker-mailserver/pull/3531))
  - Revised the "Update and Cleanup" maintenance page ([#3539](https://github.com/docker-mailserver/docker-mailserver/pull/3539), [#3583](https://github.com/docker-mailserver/docker-mailserver/pull/3583))
  - Fixed path to `rspamd.log` ([#3585](https://github.com/docker-mailserver/docker-mailserver/pull/3585))
  - Improved debugging page advice on working with logs ([#3626](https://github.com/docker-mailserver/docker-mailserver/pull/3626), [#3640](https://github.com/docker-mailserver/docker-mailserver/pull/3640))
  - "Optional Config" page now uses consistent lowercase convention for directory names ([#3629](https://github.com/docker-mailserver/docker-mailserver/pull/3629))
  - Clarified the default for ENV `FETCHMAIL_PARALLEL` ([#3603](https://github.com/docker-mailserver/docker-mailserver/pull/3603))
  - `CONTRIBUTORS.md`: Removed redundant "All Contributors" section ([#3638](https://github.com/docker-mailserver/docker-mailserver/pull/3638))
- **Tests:**
  - Refactored LDAP tests to current conventions ([#3483](https://github.com/docker-mailserver/docker-mailserver/pull/3483))
  - Changed OpenLDAP image to `bitnami/openldap` ([#3494](https://github.com/docker-mailserver/docker-mailserver/pull/3494))
  - Revised LDAP config + setup ([#3514](https://github.com/docker-mailserver/docker-mailserver/pull/3514))
  - Added tests for the helper function `_add_to_or_update_postfix_main()` ([#3505](https://github.com/docker-mailserver/docker-mailserver/pull/3505))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -21,11 +28,87 @@ All notable changes to this project will be documented in this file. The format
...
+- clarify that the issue tracker is not for personal support ([#3498](https://github.com/docker-mailserver/docker-mailserver/pull/3498))
+- change `setup config dkim` default key size to `2048` (`open-dkim`) ([#3508](https://github.com/docker-mailserver/docker-mailserver/pull/3508))
+- LDAP config improvements ([#3522](https://github.com/docker-mailserver/docker-mailserver/pull/3522))
+- ensure files are committed with `eol=lf` via `.gitattributes` ([#3527](https://github.com/docker-mailserver/docker-mailserver/pull/3527))
+- drop special bits from Postfix `maildrop/` and `public/` directory permissions ([#3625](https://github.com/docker-mailserver/docker-mailserver/pull/3625))
```
Comment (polarathene) [inline CHANGELOG.md:94]:
```suggestion
- Changed `setup config dkim` default key size to `2048` (`open-dkim`) ([#3508](https://github.com/docker-mailserver/docker-mailserver/pull/3508))
- LDAP config improvements ([#3522](https://github.com/docker-mailserver/docker-mailserver/pull/3522))
- `.gitattributes` now ensures files are committed with `eol=lf` ([#3527](https://github.com/docker-mailserver/docker-mailserver/pull/3527))
- Dropped special bits from Postfix `maildrop/` and `public/` directory permissions ([#3625](https://github.com/docker-mailserver/docker-mailserver/pull/3625))
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,11 +28,87 @@ All notable changes to this project will be documented in this file. The format
...
+- do not register `_setup_spam_to_junk()` when `SMTP_ONLY=1` ([#3385](https://github.com/docker-mailserver/docker-mailserver/pull/3385))
+- fixed issue with concatenating `$dmarc_milter` and `$dkim_milter` in `main.cf` ([#3380](https://github.com/docker-mailserver/docker-mailserver/pull/3380))
+- fixed Rspamd DKIM signing for inbound emails ([#3439](https://github.com/docker-mailserver/docker-mailserver/pull/3439), [#3453](https://github.com/docker-mailserver/docker-mailserver/pull/3453))
+- improvements to LDIF test data ([#3506](https://github.com/docker-mailserver/docker-mailserver/pull/3506))
+- tests
+  - run `pgrep` within the actual container ([#3553](https://github.com/docker-mailserver/docker-mailserver/pull/3553))
+  - `lmtp_ip.bats` improve partial failure output ([#3552](https://github.com/docker-mailserver/docker-mailserver/pull/3552))
+  - normalize for `.gitattributes` + improve `eclint` coverage ([#3566](https://github.com/docker-mailserver/docker-mailserver/pull/3566))
+- Open-DKIM key generation was broken when Rspamd was also enabled ([#3535](https://github.com/docker-mailserver/docker-mailserver/pull/3535))
```
Comment (polarathene) [inline CHANGELOG.md:111]:
```suggestion
- Miscellaneous spelling and wording improvements ([#3324](https://github.com/docker-mailserver/docker-mailserver/pull/3324), [#3330](https://github.com/docker-mailserver/docker-mailserver/pull/3330), [#3337](https://github.com/docker-mailserver/docker-mailserver/pull/3337), [#3339](https://github.com/docker-mailserver/docker-mailserver/pull/3339), [#3344](https://github.com/docker-mailserver/docker-mailserver/pull/3344), [#3367](https://github.com/docker-mailserver/docker-mailserver/pull/3367), [#3411](https://github.com/docker-mailserver/docker-mailserver/pull/3411), [#3443](https://github.com/docker-mailserver/docker-mailserver/pull/3443))
- Fixed scheduled build permissions ([#3345](https://github.com/docker-mailserver/docker-mailserver/pull/3345))
- DB now properly filters user and alias entries ([#3359](https://github.com/docker-mailserver/docker-mailserver/pull/3359))
- Fixed ShellCheck linting for BATS tests ([#3347](https://github.com/docker-mailserver/docker-mailserver/pull/3347))
- Compile Dovecot `fts_xapian` from source to match Dovecot ABI ([#3373](https://github.com/docker-mailserver/docker-mailserver/pull/3373))
- `_setup_spam_to_junk()` no longer registered when `SMTP_ONLY=1` ([#3385](https://github.com/docker-mailserver/docker-mailserver/pull/3385))
- Fixed issue with concatenating `$dmarc_milter` and `$dkim_milter` in `main.cf` ([#3380](https://github.com/docker-mailserver/docker-mailserver/pull/3380))
- Fixed Rspamd DKIM signing for inbound emails ([#3439](https://github.com/docker-mailserver/docker-mailserver/pull/3439), [#3453](https://github.com/docker-mailserver/docker-mailserver/pull/3453))
- Improvements to LDIF test data ([#3506](https://github.com/docker-mailserver/docker-mailserver/pull/3506))
- **Tests:**
  - Run `pgrep` within the actual container ([#3553](https://github.com/docker-mailserver/docker-mailserver/pull/3553))
  - `lmtp_ip.bats` improved partial failure output ([#3552](https://github.com/docker-mailserver/docker-mailserver/pull/3552))
  - Normalized for `.gitattributes` + improved `eclint` coverage ([#3566](https://github.com/docker-mailserver/docker-mailserver/pull/3566))
- Open-DKIM key generation was broken when Rspamd was also enabled ([#3535](https://github.com/docker-mailserver/docker-mailserver/pull/3535))
```

**Review by polarathene — CHANGES_REQUESTED** [review]
Revisions, I'll apply them 👍

**Review by polarathene — DISMISSED** [review]
I don't have the energy to organize the changelog to better highlight more noteworthy changes, but big kudos to @georglauterbach for taking the time to piece this all together 🙏

**Review by casperklein — APPROVED** [review]

**Comment by polarathene** [comment]
These two docs updates might be worth squeezing in last minute for this release?
- https://github.com/docker-mailserver/docker-mailserver/pull/3651
- https://github.com/docker-mailserver/docker-mailserver/pull/3654

If my one is approved before @georglauterbach responds, I'll merge both and update the changelog here 👍

**Comment by georglauterbach** [comment]
@polarathene please update this PR (Changelog) and merge. Keep the feature-freeze labels though. I will publish the tag afterwards after I will have updated the release description.

**Review by polarathene — APPROVED** [review]
I've added the 3 additional PRs, and had some time to better organize the changes for users to sift through 👍

**Comment by casperklein** [comment]
> @polarathene please update this PR (Changelog) and merge. Keep the feature-freeze labels though. I will publish the tag afterwards after I will have updated the release description.

That's bad. A new tag/release should be created directly after merging this PR. At the moment, the update check notifies about the v13 release, however there isn't a new image to download yet.

**Comment by polarathene** [comment]
> At the moment, the update-check notifies about the v13 release, however there isn't a new image to download yet.

Sorry, I wasn't aware of that and just following instructions 😓 

---

It might be better if we query the github releases for a new release instead?

```console
# `-H` => Request JSON response
# `-L` => `releases/latest` is a redirect (unlike direct release tags)
$ curl -LH "accept: application/json" https://github.com/docker-mailserver/docker-mailserver/releases/latest

{"id":42,"tag_name":"v12.1.0","update_url":"...","update_authenticity_token":"...","delete_url":"...","delete_authenticity_token":"...","edit_url":"..."}
```

With that we can probably use `tag_name`, since it's JSON just throw that into `jq`:

```console
$ curl -sLH "accept: application/json" https://github.com/docker-mailserver/docker-mailserver/releases/latest | jq -r .tag_name
v12.1.0
```

Now you have the same value for `LATEST` to handle here?:

https://github.com/docker-mailserver/docker-mailserver/blob/b663e10841d7d95f25d51efb7696c0c5b588d55f/target/scripts/update-check.sh#L19-L20

---

Although if we ever do make releases that aren't sequential (_eg: a `v12.2.0`, or more realistically a `12.1.1` security patch_), that perhaps won't work as well.

If you query the next patch/minor/major (`releases/v13.0.1`, `releases/v13.1.0`, `releases/v14.0.0`) with 3 requests instead, you'd have something a bit more resilient to that concern.

**Comment by iruizr7** [comment]
Big thanks to everyone involved! This project is definitely the easiest and quickest way of installing a personal or SOHO email server with all the guarantees. Keep on team! :fireworks:

**Comment by casperklein** [comment]
> It might be better if we query the github releases for a new release instead?

Afaik the API is rate limited. If there are other things running on the same host using the API, that can lead to problems. I would stick with the current approach. Also it's rather unusual to merge the release PR, but not creating a release 😉

**Comment by georglauterbach** [comment]
I will take care within the next hours.

**Comment by georglauterbach** [comment]
That commit message though:

image

😆😆😆😆😆

**Comment by polarathene** [comment]
> Afaik the API is rate limited.

What rate limit?

- You're probably thinking of the API service at [`https://api.github.com`](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28)?
- If it were rate-limited like the actual GH REST API, it would [include that information in the response](https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api?apiVersion=2022-11-28#using-the-response)?
- Even then it's [60 requests per hour](https://docs.github.com/en/rest/overview/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-unauthenticated-users), if the DMS host were to exceed that from other activity it's likely the admin is experiencing that problem elsewhere too.

However, in this case it's just a standard web request to Github. It is effectively **no different of a request than the one you're already making** for a file at the repo?:

https://github.com/docker-mailserver/docker-mailserver/blob/2c602299136fcac8a8fd02ada86039687e6ae18c/target/scripts/update-check.sh#L7

Which you default to **once a day**:

https://github.com/docker-mailserver/docker-mailserver/blob/2c602299136fcac8a8fd02ada86039687e6ae18c/target/scripts/update-check.sh#L15

---

> I would stick with the current approach.

Your approach is what caused the false-positive for a new release being available. The approach I've suggested doesn't have any notable regression but avoids the problem scenario you noticed from occurring?

Seems like it'd be better to query the actual latest GH release and compare that to the `VERSION`.

`VERSION` may be better suited as an ENV `DMS_RELEASE` in the `Dockerfile` (_which could additionally be provided as an `ARG` since the CI should have the tag on release, avoiding a past issue where `VERSION` was not incremented with a release tag_).

> Also it's rather unusual to merge the release PR, but not creating a release 😉

Already stated that I was following @georglauterbach request. If the version check is handled the way I've proposed it fixes two release bugs related to it.

It shouldn't matter if there is a delay between a release PR being merged and when a release is officially tagged/published. This had a less than 15hr delay which I have no issue with, nothing else was being merged inbetween.

**Comment by georglauterbach** [comment]
It is definitely my fault that there is a delay here between the merging of this PR and the release; I apologize for that - it won't happen again.

**Comment by casperklein** [comment]
> You're probably thinking of the API service at https://api.github.com?
> If it were rate-limited like the actual GH REST API, it would include that information in the response?

Yes, I've things mixed up 👍 

My main point was just, that the VERSION file shouldn't be updated, when not making a new release (makes no sense). Just do both at the same time and everything is good. Fixing problems (enhancing the update check) that aren't real problems seemed unnecessary to me.

That said, it makes no difference to me checking the version file or the latest release as you suggested. Normally the results shouldn't differ. If you however like to change the check, go for it 👍

> Already stated that I was following @georglauterbach request.

I didn't want to offense you or Georg in any way. I just wanted to make aware of this and let Georg know, not to delay the release for too long. It's really no big issue to make the release some hours later. We are humans after all.

> I apologize for that - it won't happen again.

Again, if I sounded too harsh, that wasn't my intention! ✌️

**Comment by polarathene** [comment]
> My main point was just, that the VERSION file shouldn't be updated, when not making a new release (makes no sense). Just do both at the same time and everything is good.

That can be automated:

<details>
<summary>Verbose breakdown</summary>

- The published tag has the same content as `VERSION` would, so reference that in the CI (_see example from the [docs deploy workflow](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/.github/workflows/docs-production-deploy.yml#L33-L41)_). `github.ref` should be sufficient (_like [used here](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/.github/workflows/generic_publish.yml#L79)_).
- Then just add an `ARG` in the final `Dockerfile` stage, with a related `ENV` at the end (_to minimize [cache invalidation](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/Dockerfile#L327-L330)_). An `ARG` can have a [default for builds that aren't published](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/Dockerfile#L299).

`github.ref` will differ for publishing `:edge`, so we'll still need to handle that.

The [`github` context object docs](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) suggest that we could use `github.ref_name` to get just the tag value and `github.ref_type` to determine if it's a tag trigger with a [conditional expression](https://docs.github.com/en/actions/learn-github-actions/expressions#example).

We could add the expression [here to set `DMS_RELEASE` as a new `ARG` input](https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/.github/workflows/generic_publish.yml#L78-L80).

---

</details>

<details>
<summary>Reference - Sources</summary>

https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/.github/workflows/generic_publish.yml#L78-L80

https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/Dockerfile#L297-L299

https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/Dockerfile#L327-L330

https://github.com/docker-mailserver/docker-mailserver/blob/b663e10841d7d95f25d51efb7696c0c5b588d55f/target/scripts/update-check.sh#L6

https://github.com/docker-mailserver/docker-mailserver/blob/68a43eb4970f2ab7680ffa470a1d26e37fa375f0/target/scripts/start-mailserver.sh#L123

</details>

```yaml
          build-args: |
            DMS_RELEASE=${{ github.ref_type == 'tag' && github.ref_name || 'edge' }}
            VCS_REVISION=${{ github.sha }}
````

```Dockerfile
FROM stage-main AS stage-final
ARG DMS_RELEASE=edge
ARG VCS_REVISION=unknown
```

```Dockerfile
# ARG invalidates cache when it is used by a layer (implicitly affects RUN)
# Thus to maximize cache, keep these lines last:
LABEL org.opencontainers.image.revision=${VCS_REVISION}
LABEL org.opencontainers.image.version=${DMS_RELEASE}
ENV DMS_RELEASE=${DMS_RELEASE}
```

```bash
VERSION="${DMS_VERSION}"
```

```bash
  [[ ${ENABLE_UPDATE_CHECK} -eq 1 && ${DMS_RELEASE} != 'edge' ]] && _register_start_daemon '_start_daemon_update_check'
```

You could of course create the `/VERSION` file instead of an ENV this way, but with an ENV it's more clear that the service isn't used with `:edge` (_doesn't really make sense to?_)

---

> Fixing problems (enhancing the update check) that aren't real problems seemed unnecessary to me.

1. `:edge` should not care about the update process, these concerns should not have been raised:
   - https://github.com/docker-mailserver/docker-mailserver/pull/3659
   - https://github.com/docker-mailserver/docker-mailserver/pull/3662#pullrequestreview-1749456998
2. In the past a release was made [without bumping `VERSION` accordingly](https://github.com/docker-mailserver/docker-mailserver/issues/2952#issuecomment-1364304849).
   - The fix was to [publish a new patch version/tag](https://github.com/docker-mailserver/docker-mailserver/pull/2944).
   - Wouldn't be a problem with improved update check logic.
3. This concern you raised with the release PR and delayed tagging of the release would again be a non-issue.
   - `:edge` unaffected since it's not actually tied to the `/VERSION` file it compares to Github `VERSION` with.
   - Tagged releases wouldn't notify until a new tagged release is actually published.

All concerns related to the update checking script, all resolved by better integration + checking.

---

> Normally the results shouldn't differ. If you however like to change the check, go for it 👍

Agreed, I'll raise a PR so these concerns don't occur again.

> Edit: After more thinking, I came to the conclusion, that checking the release is more logical and therefore better.

🎉 

> I didn't want to offense you or Georg in any way. I just wanted to make aware of this and let Georg know, not to delay the release for too long. It's really no big issue to make the release some hours later. We are humans after all.

No offense taken! 😄

**Pull request merged**